### PR TITLE
feat(mantle): keep track of Blend-related state for epoch-bound PoW difficulty updates

### PR DESCRIPTION
**Pull request by ntn-x2**
## 1. What does this PR implement?

PR built on top of the PoW Blend proof generators (https://github.com/logos-blockchain/logos-blockchain/pull/3310). It introduces the mechanism, defined in the [EmPoWering spec](https://github.com/logos-co/logos-lips/pull/400), to update the Blend puzzle difficulty for PoW.

The spec contains all the math, but in short: we keep track of tx density on each new block, inside the PoW ledger, and at the time we take the epoch nonce snapshot, we compute also the new difficulty for that epoch.

While working on this I noticed a few things that need be fixed or can be improved, and I will open some PRs after this whole PoW feature is implemented on Blend side.

## 2. Does the code have enough context to be clearly understood?

Yes.

## 3. Who are the specification authors and who is accountable for this PR?

@madxor spec, @ntn-x2 impl

## 4. Is the specification accurate and complete?

Yes, so far.

## 5. Does the implementation introduce changes in the specification?

Changes were requested and applied during the implementation phase. At this point the PR is in sync with the implementation.

### CODE REVIEW

File: deployment/ceremony/genesis/devnet/deployment-template.yaml [hunk 1]
```diff
@@ -27,6 +27,16 @@ cryptarchia:
     numerator: 1
     denominator: 30
   learning_rate: 1.0
+  pow_config:
+    blend:
+      # BLEND_DIFFICULTY_BASE, as the exponent n in p / 2^n. The baseline
+      # is TBD in the spec pending a measurement on the target hardware;
+      # 19 is the value it currently gives.
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply deployment/ceremony/genesis/devnet/deployment-template.yaml]: Fixed in https://github.com/logos-blockchain/logos-blockchain/pull/3317/commits/942e6112a688416a24227027d1ede30477b9154e.

File: deployment/ceremony/genesis/standalone/deployment-template.yaml [hunk 2]
```diff
@@ -27,6 +27,16 @@ cryptarchia:
     numerator: 1
     denominator: 20
   learning_rate: 0.5
+  pow_config:
+    blend:
+      # BLEND_DIFFICULTY_BASE, as the exponent n in p / 2^n. The baseline
+      # is TBD in the spec pending a measurement on the target hardware;
+      # 19 is the value it currently gives.
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply deployment/ceremony/genesis/standalone/deployment-template.yaml]: Fixed in https://github.com/logos-blockchain/logos-blockchain/pull/3317/commits/942e6112a688416a24227027d1ede30477b9154e.

File: deployment/ceremony/genesis/testnet/deployment-template.yaml [hunk 3]
```diff
@@ -27,6 +27,16 @@ cryptarchia:
     numerator: 1
     denominator: 30
   learning_rate: 1.0
+  pow_config:
+    blend:
+      # BLEND_DIFFICULTY_BASE, as the exponent n in p / 2^n. The baseline
+      # is TBD in the spec pending a measurement on the target hardware;
+      # 19 is the value it currently gives.
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply deployment/ceremony/genesis/testnet/deployment-template.yaml]: Fixed in https://github.com/logos-blockchain/logos-blockchain/pull/3317/commits/942e6112a688416a24227027d1ede30477b9154e.

File: nodes/node/binary/src/config/deployment/settings.yaml [hunk 4]
```diff
@@ -27,6 +27,16 @@ cryptarchia:
     numerator: 1
     denominator: 20
   learning_rate: 0.5
+  pow_config:
+    blend:
+      # BLEND_DIFFICULTY_BASE, as the exponent n in p / 2^n. The baseline
+      # is TBD in the spec pending a measurement on the target hardware;
+      # 19 is the value it currently gives.
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply nodes/node/binary/src/config/deployment/settings.yaml]: Fixed in https://github.com/logos-blockchain/logos-blockchain/pull/3317/commits/942e6112a688416a24227027d1ede30477b9154e.

File: nodes/node/standalone-deployment-config.yaml [hunk 5]
```diff
@@ -27,6 +27,16 @@ cryptarchia:
     numerator: 1
     denominator: 20
   learning_rate: 0.5
+  pow_config:
+    blend:
+      # BLEND_DIFFICULTY_BASE, as the exponent n in p / 2^n. The baseline
+      # is TBD in the spec pending a measurement on the target hardware;
+      # 19 is the value it currently gives.
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply nodes/node/standalone-deployment-config.yaml]: Fixed in https://github.com/logos-blockchain/logos-blockchain/pull/3317/commits/942e6112a688416a24227027d1ede30477b9154e.

File: ledger/src/mantle/sdp/rewards/blend/mod.rs [hunk 6]
```diff
@@ -258,19 +261,30 @@ impl RewardsParameters {
...
             lottery_1: epoch_state.lottery_1,
         }
     }
+
+    /// The epoch's proof of work public inputs: the threshold the chain fixed
+    /// for this epoch, and `Q_W = ß_max`, so one solution buys one message.
+    fn pow_inputs(&self, epoch_state: &EpochState) -> PowInputs {
+        PowInputs {
+            pow_blend_difficulty: epoch_state.blend_pow_difficulty,
+            pow_quota: self.encapsulations_per_message(),
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply ledger/src/mantle/sdp/rewards/blend/mod.rs]: Fixed in https://github.com/logos-blockchain/logos-blockchain/pull/3317/commits/942e6112a688416a24227027d1ede30477b9154e. There is a similar issue with the leadership quota calculation. That will also tackled in a different PR.

File: ledger/src/config.rs [hunk 7]
```diff
@@ -92,6 +95,43 @@ impl Config {
...
+impl BlendPoWConfig {
+    /// The damping exponent `alpha = a / b <= 1`, as the pair `(a, b)`.
+    ///
+    /// Both are exponents applied to big integers, so they must stay small —
+    /// `alpha` is a simple fraction such as `1/2`, not a high-precision ratio.
+    #[must_use]
+    pub const fn damping_exponent(&self) -> (u32, NonZeroU32) {
+        let numerator = self.damping_num.get();
+        let denominator = NonZeroU32::new(numerator.strict_add(self.damping_den_offset))
+            .expect("Numerator is non-zero, so denominator must be non-zero as well since it's a non-negative offset from the numerator.");
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply ledger/src/config.rs:130]: I will introduce a new `BoundedInteger` type that we can use to enforce all these constraint statically. For now I will not address this comment.

File: services/blend/src/core/mod.rs [hunk 8]
```diff
@@ -1017,6 +1022,10 @@ where
                 core_poq_generator: new_core_poq_generator,
                 public: new_epoch_info,
             } = *core_epoch_info;
+            // Once a new epoch starts, old epoch's PoW work is useless, so we drop the PoW
+            // proof generator for the epoch transition period.
+            let mut current_cryptographic_processor = current_cryptographic_processor;
+            current_cryptographic_processor.stop_proof_generation();
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply services/blend/src/core/mod.rs:1028]: Fixed in https://github.com/logos-blockchain/logos-blockchain/pull/3317/commits/942e6112a688416a24227027d1ede30477b9154e.

File: ledger/src/cryptarchia/stake.rs [hunk 9]
```diff
@@ -200,6 +201,15 @@ mod tests {
                 },
             },
             faucet_pk: None,
+            pow_config: crate::config::PoWConfig {
+                blend: crate::config::BlendPoWConfig {
```
Comment (danielSanchezQ) [inline ledger/src/cryptarchia/stake.rs:205]: lets import stuff and remove the qualified paths

File: ledger/src/mantle/pow/mod.rs [hunk 10]
```diff
@@ -158,15 +192,29 @@ impl PowState {
...
+
+    /// The load of the last epoch to close — the observation the Blend
+    /// difficulty retarget reads — or `None` while no epoch has closed yet.
+    pub(crate) const fn last_closed_epoch_load(&self) -> Option<ClosedEpochLoad> {
+        self.blend.tx_density.last_closed_epoch_load()
+    }
+
+    /// Apply an epoch transition: on epoch change, refill the reward pool from
+    /// the rewards collected during `previous_epoch`, and close that epoch's
+    /// transaction totals so the Blend retarget can read them.
```
Comment (danielSanchezQ) [inline ledger/src/mantle/pow/mod.rs:217]: AI just added blend related content, this doesn't reflect on the methods usage iirc.

File: ledger/src/mantle/pow/tx_density.rs [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+pub struct TxDensity {
+    /// Totals of the epoch the ledger is currently extending, still growing
+    /// with every applied block.
+    current_epoch: EpochTotals,
+    /// Totals of the most recently closed epoch, final and no longer moving.
+    ///
+    /// `None` until the first epoch closes: before that no complete epoch
+    /// exists to read a load from, which is a different state from an epoch
+    /// that closed having carried nothing.
+    last_closed_epoch: Option<EpochTotals>,
```
Comment (danielSanchezQ) [inline ledger/src/mantle/pow/tx_density.rs:24]: Nitpick: we could have use a different pattern here. If the density has two states depending if there was a closed epoch or not it could be encapsulated. Setting up optionals is usually a burden and may lead to miss setting it up. 
To be fair this is easy enough, so lets just go with it.

File: ledger/src/mantle/pow/tx_density.rs [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
+struct EpochTotals {
+    blocks: u64,
+    txs: u64,
+}
+
+impl EpochTotals {
+    const fn record_block(&mut self, txs_in_block: u64) {
+        self.blocks = self.blocks.saturating_add(1);
+        self.txs = self.txs.saturating_add(txs_in_block);
```
Comment (danielSanchezQ) [inline ledger/src/mantle/pow/tx_density.rs]: I'm not sure saturating is the way to go here. Let me know if this logic makes sense:
Overflowing a `u64` on blocks produced in a single epoch is virtually impossible. Same with txs (probably?)
Saturating may be masking a serious problem, should we just panic and stop computing? 🤔

File: ledger/src/config.rs [hunk 13]
```diff
@@ -92,6 +95,43 @@ impl Config {
     }
 }
 
+#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, PartialEq, Eq)]
+// TODO: Add reward difficulty parameters here. For now only the ones used for
+// Blend are included.
```
Comment (danielSanchezQ) [inline ledger/src/config.rs:100]: Lets open an issue for this as well. I should have done it in the first place anyway. Thanks.

File: zk/groth16/src/lib.rs [hunk 14]
```diff
@@ -68,10 +70,23 @@ pub fn fr_from_bytes(fr: &[u8]) -> Result<Fr, impl Error + use<>> {
...
-    let n: BigUint = n % BigUint::from(<Fr as PrimeField>::MODULUS);
+    let n: BigUint = n % fr_modulus();
     n.into()
 }
 
+/// The scalar field modulus `p`.
+#[must_use]
+pub fn fr_modulus() -> BigUint {
+    Fr::MODULUS.into()
+}
```
Comment (danielSanchezQ) [inline zk/groth16/src/lib.rs:81]: Imo, this is worst to read than the old code.

File: zk/groth16/src/modulus_shift.rs [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    #[must_use]
+    pub const fn new<const SHIFT: u32>() -> Self {
+        const {
+            assert!(
+                SHIFT >= 1 && SHIFT <= Self::MAX,
+                "shift does not denote a field element"
+            );
+        }
+        Self(SHIFT)
+    }
```
Comment (danielSanchezQ) [inline zk/groth16/src/modulus_shift.rs:44]: Nice stuff!!

**Review by youngjoon-lee — APPROVED** [review]
I mainly focused on the `lb-ledger`. The logic looks correct.

> While working on this I noticed a few things that need be fixed or can be improved, and I will open some PRs after this whole PoW feature is implemented on Blend side.

FYI: If you're thinking about polishing the ledger code, I'm also working on refactoring the ledger to make epoch transition/snapshot much easier to maintain. I'm going to open PRs at the next iteration, to not block any new features to be implemented in the current iteration.

File: ledger/src/cryptarchia/stake.rs (hunk 9 above)
Comment (ntn-x2) [reply ledger/src/cryptarchia/stake.rs:205]: We're using full imports everywhere else in that file, so I wanted to keep things consistent.

File: ledger/src/mantle/pow/mod.rs (hunk 10 above)
Comment (ntn-x2) [reply ledger/src/mantle/pow/mod.rs:217]: What do you mean? This is part of the PR, we do a bunch of computations for both rewards and Blend side of the PoW machinery. We calculate the total tx density so we can compute the new Blend difficulty.

File: ledger/src/mantle/pow/tx_density.rs (hunk 12 above)
Comment (ntn-x2) [reply ledger/src/mantle/pow/tx_density.rs]: I mean, it's really unlikely this value will reach its maximum. But even if it does, I think in this case saturating is fine as long as every other node will do the same. We use the total count to derive the new Blend difficulty, so I think saturating would have the result of clamping the maximum difficulty, in already a edgy case of having so many txs and blocks, so I don't think panicking/erroring would be any better.

File: ledger/src/cryptarchia/stake.rs (hunk 9 above)
Comment (danielSanchezQ) [reply ledger/src/cryptarchia/stake.rs:205]: Damn, ok then. We can change later.

File: ledger/src/config.rs (hunk 13 above)
Comment (ntn-x2) [reply ledger/src/config.rs:100]: https://github.com/logos-blockchain/logos-blockchain/issues/3343

File: ledger/src/mantle/pow/mod.rs (hunk 10 above)
Comment (danielSanchezQ) [reply ledger/src/mantle/pow/mod.rs:217]: Uhm, I think yesterday something went wrong. That is not what was showing up. Maybe something related to gh outage. Disregard this.

File: zk/groth16/src/lib.rs (hunk 14 above)
Comment (ntn-x2) [reply zk/groth16/src/lib.rs:81]: We need the saturating function, so you mean replacing the `Fr::MODULUS` with the `fr_modulus()` function?

File: ledger/src/mantle/pow/tx_density.rs (hunk 12 above)
Comment (danielSanchezQ) [reply ledger/src/mantle/pow/tx_density.rs]: In the happy case it will not make a difference, but in the sad event of us making an error it would imo as this would be masking the error.

Comment (ntn-x2) [reply ledger/src/mantle/pow/tx_density.rs]: But what is the happy case? I think it's totally reasonable to think we could have that many transactions, and in that case what would panicking do? I think we need to consider these values going above the maximum as either a valid case or an invalid case. If it's a valid case, I don't see why we would panic. If it's not a valid case, then I agree with you, although I do think this is a valid case and would not be possible to distinguish between the value going genuinely beyond the limit or us artificially inflating it due to a coding error.

Comment (danielSanchezQ) [reply ledger/src/mantle/pow/tx_density.rs]: From the top of my head, an epoch (with current values) can have at most about 6k blocks (if every slot would fit a block which is not the case neither, but lets go with limits). We have a limit in 1024 txs per block. As this epoch state is for a single chain (and fork); this means that at most we could have around 6M txs per epoch. Making this virtually impossible to fail. Hence if it fails it falls into an error on our side, the node is missbehaving and should not continue to operate (imo).
But, lets at least take the middle ground, check for overflowing and raise a warning/error log if that happens. wdyt?

Comment (ntn-x2) [reply ledger/src/mantle/pow/tx_density.rs]: Ok that makes sense, and we then need a type for that, so we enforce everything at compile time, assuming the bounds are static and not configuration-specific. I will change it to an expect then, it makes sense now. Thanks!

File: zk/groth16/src/lib.rs (hunk 14 above)
Comment (danielSanchezQ) [reply zk/groth16/src/lib.rs:81]: Yeah, could we at least make it static in the method so we do not call the into in each call?

**Review by danielSanchezQ — APPROVED** [review]
